### PR TITLE
[WIP] Change hashing algorithm in Deployments

### PR DESCRIPTION
**Pull request by 0xmichalis**
Superseded by https://github.com/kubernetes/kubernetes/pull/44774

### CODE REVIEW

**Comment by kargakis** [comment]
@wojtek-t I want to write a test that spins up a cluster, creates X namespaces, creates Y deployments in each namespace, runs Z rollouts for each deployment, restart the controller manager with `--deployment-hashing-algorithm=migrates-to-fnv`, kick new rolllouts for all the existing deployments, and verify in the end that all replica sets have migrated. Any ideas where such a test should live?

X (average # of namespaces per cluster)
Y (average # of deployments per namespace)
Z (can be 2, can be 200)

**Comment by kargakis** [comment]
@smarterclayton where do we expose metrics for the controller manager? I can't seem to find those under the master /metrics endpoint.

**Comment by bgrant0607** [comment]
cc @janetkuo

**Comment by smarterclayton** [comment]
No idea - scheduler has it, and I thought it was enabled for controller manager because someone was looking at queue depths a while back.

**Comment by janetkuo** [comment]
> where do we expose metrics for the controller manager?

Is this what you're looking for?
https://github.com/kubernetes/kubernetes/blob/master/pkg/metrics/metrics_grabber.go#L122

```go
func (g *MetricsGrabber) GrabFromControllerManager() (ControllerManagerMetrics, error) {
	if !g.registeredMaster {
		return ControllerManagerMetrics{}, fmt.Errorf("Master's Kubelet is not registered. Skipping ControllerManager's metrics gathering.")
	}
	output, err := g.getMetricsFromPod(fmt.Sprintf("%v-%v", "kube-controller-manager", g.masterName), api.NamespaceSystem, ports.ControllerManagerPort)
	if err != nil {
		return ControllerManagerMetrics{}, err
	}
	return parseControllerManagerMetrics(output)
}
```

**Comment by kargakis** [comment]
@janetkuo the library you are linking is assuming we are running a self-hosted cluster (components run inside pods). I can't see us using it anywhere in the Kubernetes repository apart from some e2e tests.

File: pkg/apis/componentconfig/types.go [hunk 1]
```diff
@@ -789,6 +789,10 @@ type KubeControllerManagerConfiguration struct {
 	// ReconcilerSyncLoopPeriod is the amount of time the reconciler sync states loop
 	// wait between successive executions. Is set to 5 sec by default.
 	ReconcilerSyncLoopPeriod metav1.Duration
+	// deploymentHashingAlgorithm is the hasing algorithm used for hashing pod templates for Deployments.
```
Comment (janetkuo) [inline pkg/apis/componentconfig/types.go]: typo: has**h**ing

File: pkg/controller/deployment/deployment_controller.go [hunk 2]
```diff
@@ -27,7 +27,9 @@ import (
 	"time"
 
 	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
 	"k8s.io/apimachinery/pkg/api/errors"
+
```
Comment (janetkuo) [inline pkg/controller/deployment/deployment_controller.go]: remove newline

File: pkg/controller/deployment/sync.go [hunk 3]
```diff
@@ -164,7 +165,26 @@ func (dc *DeploymentController) addHashKeyToRSAndPods(rs *extensions.ReplicaSet)
...
+		hash = deploymentutil.GetReplicaSetHashFnv(rs)
+	case "migrate-to-fnv":
+		// TODO: Should we look for the migrated-from annotation?
+		// We can look at the name suffix of the replica set, if it's the same as
+		// the old hash, then use the old hash, otherwise use the new algorithm.
+		if !strings.HasSuffix(rs.Name, hash) {
+			hash = deploymentutil.GetReplicaSetHashFnv(rs)
+		} else {
+			hash = deploymentutil.GetReplicaSetHash(rs)
+		}
```
Comment (janetkuo) [inline pkg/controller/deployment/sync.go]: Should we handle the default case (return error)?

**Comment by janetkuo** [comment]
@kargakis FYI the controller manager metrics handler is [here](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-controller-manager/app/controllermanager.go#L140)

**Comment by kargakis** [comment]
@janetkuo thanks for the pointer.

Note that if the update on the Deployment proposal that I have up for review is accepted (https://github.com/kubernetes/community/pull/384), then we may not need to migrate old replica sets.

**Comment by kargakis** [comment]
A migration won't be needed. Closing this one

**Pull request closed without merging**